### ISSUE TITLE
phpspec not releasing memory/using a lot of memory

### ISSUE THREAD

**Issue by dstockto** [phpspec/phpspec#884]
We've got ~1300 tests in our suite currently and just ran into an issue where PHP was unable to allocate memory because it had hit the 128M limit we have set. Of course we can bump that up, but it doesn't seem to make sense that it would need to keep all the objects around after the tests have run.

The error that happened that made us aware of this was:

```
Fatal error happened while executing the following 
    it will build the ObjectStatusMap service 
    Allowed memory size of 134217728 bytes exhausted (tried to allocate 72 bytes) in /vagrant_web/vendor/phpspec/phpspec/src/PhpSpec/Listener/StatisticsCollector.php on line 89 
```

Interestingly, if we run with format=pretty, it succeeds. If we comment out some tests, it also works. 

I'm looking at potentially parallelizing the run so fewer tests will run within a single run of phpspec, but it still leaves us in the same state if the individual namespaces grow to a similar number/size of tests as the full suite.

**Comment by everzet** [comment]
Sounds like progress formatter has a memory leak somewhere.

**Comment by dstockto** [comment]
Sorry, I should have mentioned the formatter we normally use is dot.

**Comment by shanethehat** [comment]
I think this is caused by the dot formatter getting a count of all the events in the suite. It does this by performing an array_merge on the existing arrays of events and then counting the resulting bigger array. That seems a bit wasteful given that it's just duplicating stored data.

**Comment by shanethehat** [comment]
@dstockto would you mind testing if that PR solves your issue?

**Comment by stof** [comment]
@shanethehat giving a link to the PR would help

**Comment by shanethehat** [comment]
Indeed, edited.

**Comment by dstockto** [comment]
Here's my results. I don't remember what we changed that made this be less of an issue, but I lowered my memory limit until I got it to fail. In this case, it was 77M for my current test suite without the code from the patch applied. At 78M it succeeds, at 77M, I get 
```
Fatal error happened while executing the following
     [exec]     it will not allow deleteList
     [exec]     Allowed memory size of 80740352 bytes exhausted (tried to allocate 32768 bytes) in /vagrant_web/vendor/phpspec/prophecy/src/Prophecy/Doubler/Generator/ClassCreator.php(49) : eval()'d code on line 
```

With the patch applied, unfortunately, I get the exact same result. It fails on the same line of code asking for the same amount of memory. With 78M it passes.

**Comment by shanethehat** [comment]
Hmm.... it's not the same file as last time :disappointed:

**Comment by dstockto** [comment]
No, unfortunately. I don't have the same situation I did before. We've changed some things, deleted some modules and thus some of the tests, so it's not blowing up with a 128M limit any more. If I get some time tomorrow or soon, perhaps I can do a git bisect to find the spot where it failing on 128M and see if the patch works there. It's much too late this evening, though.

**Comment by ciaranmcnulty** [comment]
@dstockto If you can look at @shanethehat's PR with your bigger suite it would be really useful. Thanks!

**Comment by Aeolun** [comment]
@ciaranmcnulty @shanethehat I have this same issue with our suite, and regardless of whether I apply the change in the pull request, I still have the suite failing. It unfortunately fails at the same point too.

`Allowed memory size of 134217728 bytes exhausted (tried to allocate 72 bytes)`

**Comment by gui-don** [comment]
Same problem here. It seems unrelated to the formatter for me. Each test (`ExampleRunner::executeExample`) adds between 50ko and 300ko to the total memory usage.

**Comment by naomigo** [comment]
I am also running into the memory issue, with approx 400 tests in 32 spec files:

 `Allowed memory size of 134217728 bytes exhausted (tried to allocate 32 bytes)`

I have one spec class that has approx 70 tests in it which takes 19 seconds to run through (a lot longer than any of the other specs). Wondering if it's because I could be mocking too many objects?

**Comment by naomigo** [comment]
Some further information on my spec class that is taking 20s to run - I suspect the let() method might be doing too much?

`<?php

namespace spec\Application\Model;

use PhpSpec\ObjectBehavior;
use Prophecy\Argument;

class MyClassSpec extends ObjectBehavior
{
	private $diMock;
	
	private $class1Mock;
	private $class2Mock;	
	private $class3Mock;
	private $class4Mock;
	private $class5Mock;	
	private $class6Mock;
	private $class7Mock;
	private $class8Mock;	
	private $class9Mock;	
	private $class10Mock;
	private $class11Mock;	
	private $class12Mock;
	private $class13Mock;		
	private $class14Mock;
	private $class15Mock;
	private $class16Mock;
	
	private $value1Mock;
	private $value2Mock;
	private $value3Mock;
	private $value4Mock;
	private $value5Mock;
	private $value6Mock;
	private $value7Mock;
	
    function it_is_initializable()
    {
        $this->shouldHaveType('Application\Model\MyClass');
    }
    
    function let(\DI\Container $di, \Path\To\Class1 $class1, \Path\To\Class2 $class2,
    		\Path\To\Class3 $class3,
    		\Path\To\Class4 $class4,
    		\Path\To\Class5 $class5,
    		\Path\To\Class6 $class6,
    		\Path\To\Class7 $class7,
    		\Path\To\Class8 $class8,
    		\Path\To\Class9 $class9,
    		\Path\To\Class10 $class10,
    		\Path\To\Class11 $class11,
    		\Path\To\Class12 $class12,
    		\Path\To\Class13 $class13,
    		\Path\To\Class14 $class14,
    		\Path\To\Class15 $class15,
    		\Path\To\Class16 $class16)
    {
    	$this->diMock = $di;
    	
    	$this->diMock->get('class1')->shouldBeCalled()->willReturn($class1);
    	$this->class1Mock = $class1;    	
    	$this->class2Mock = $class2;

    	$this->diMock->get('class3')->willReturn($class3);
    	$this->class3Mock = $class3;    	
    	$this->diMock->get('class4')->willReturn($class4);
    	$this->class4Mock = $class4;    	
    	$this->diMock->get('class5')->willReturn($class5);
    	$this->class5Mock = $class5;   	

    	$this->diMock->get('class6')->shouldBeCalled()->willReturn($class6);
    	$this->class6Mock = $class6;
    	$this->diMock->get('class7')->shouldBeCalled()->willReturn($class7);
    	$this->class7Mock = $class7;
    	$this->diMock->get('class8')->shouldBeCalled()->willReturn($class8);
    	$this->class8Mock = $class8;    	 
    	$this->diMock->get('class9')->shouldBeCalled()->willReturn($class9);
    	$this->class9Mock = $class9;    	
    	
    	$this->diMock->get('class10')->willReturn($class10);
    	$this->class10Mock = $class10;    	
    	$this->diMock->get('class11')->willReturn($class11);
    	$this->class11Mock = $class11;    	
    	
    	$this->diMock->get('class12')->shouldBeCalled()->willReturn($class12);
    	$this->class12Mock = $class12;
    	$this->diMock->get('class13')->shouldBeCalled()->willReturn($class13);
    	$this->class13Mock = $class13;    	
    	
    	$this->diMock->get('class14')->willReturn($class14);
    	$this->class14Mock = $class14;    	
    	$this->diMock->get('class15')->willReturn($class15);
    	$this->class15Mock = $class15;
    	$this->diMock->get('class16')->willReturn($class16);
    	$this->class16Mock = $class16;
    	
    	$this->diMock->get('value1')->willReturn('Value 1');
    	$this->value1Mock = 'Value 1';    	
    	$this->diMock->get('value2')->willReturn('Value 2');
    	$this->value2Mock = 'Value 2';    	
    	$this->diMock->get('value3')->willReturn('Value 3');
    	$this->value3Mock = 'Value 3';
    	$this->diMock->get('value4')->willReturn('Value 4');
    	$this->value4Mock = 'Value 4';
    	$this->diMock->get('value5')->willReturn('Value 5');
    	$this->value5Mock = 'Value 5';
    	$this->diMock->get('value6')->willReturn('Value 6');
    	$this->value6Mock = 'Value 6';
    	$this->diMock->get('value7')->willReturn('Value 7');
    	$this->value7Mock = 'Value 7';
    	
    	$this->beConstructedWith($this->diMock);
    }
}`

**Comment by jakzal** [comment]
@naomigo aside the memory issue which should be fixed in phpspec, you should probably work on your design a bit more. 

A lot of collaborators is a smell. Specs with lots of collaborators are hard to read and maintain. It's often hard to change the code driven this way too.

When you find yourself writing lots of mocks, it's time to revise the design and see how you can decrease number of collaborators. Perhaps your class is doing too much. Perhaps some collaborators could be hidden behind a facade (not the laravel one). It's hard to say what's wrong exactly without seeing the actual code/specs. Bottom line is there's too much mocking going on in your specs ;)

**Comment by stof** [comment]
Why are you storing all mocks as properties in your spec classes ? This is probably what creates issues for you, because it is very likely to create cyclic object graphs, which cannot be released from memory by the refcount (and thus stay in memory until the GC collects them).

Phpspec relies on argument names to inject the same collaborator instance in ``let`` and in the actual example.

**Comment by naomigo** [comment]
Thanks for the replies.

I guess some of the reason "why" I'm doing things is how I've learnt/been taught them. My understanding is that I'm storing mocks as properties because I can then use the property within the tests, e.g.

`function it_should_do_something()
{
    $this->class1Mock->getData()->shouldBeCalled()->willReturn(array());
}

function it_should_do_something_else()
{
    $this->class1Mock->getData()->shouldBeCalled()->willReturn(array(1,2,3));
}
`
Instead of the above, should I be doing something like:
`
function it_should_do_someting(\Path\To\Class1 $class1)
{
    $this->diMock->get('class1')->shouldBeCalled()->willReturn($class1);
    $class1->getData()->shouldBeCalled()->willReturn(array());
}
`
Perhaps I'm missing some fundamental understanding somewhere on how to mock objects?

I've had concerns about this particular class for some time, in that it needs a lot of collaborators. I think part of the reason for this is that it's functionality is similar to that of a controller class (handling the flow of a registration process) and I have read previously that sometimes it is not possible (or becomes very difficult) to spec controller classes...

**Comment by ciaranmcnulty** [comment]
@naomigo Hes the second example is how you should do it in PhpSpec. The tool will take care of maintaining the same instance between `let` and the examples.

You're right to feel concern about the class though, and the tests are highlighting the issue. Consider whether some of them are related can be grouped into objects that then delegate to the existing components.

**Comment by stof** [comment]
The right way to do it in PhpSpec is even different from your second case:

```php
function let(\DI\Container $di)
{
    // Do some initialization of the $di test double, but don't store it in a property
}

function it_should_do_something($di, \Path\To\Class1 $class1)
{
    $di->get('class1')->shouldBeCalled()->willReturn($class1);

    $class1->getData()->shouldBeCalled()->willReturn(array());
}
```

``$di`` will be injected based on what was done in ``let()`` (you can typehint it again but it is not needed).

Btw, totally unrelated to Phpspec, saying ``\Path\To\Class1`` is wrong. It is not a path at all.

**Comment by ciaranmcnulty** [comment]
It's true you don't need the typehint separately but it's slightly nicer in IDEs if you add it IMO

**Comment by stof** [comment]
@ciaranmcnulty the IDE already complain, because ``$class1->getData()`` does not return a MethodProphecy as far as the IDE knows, and so it tells you that ``shouldBeCalled()`` does not exist...

**Comment by naomigo** [comment]
Thanks for the additional comments.

Just to confirm my general phpspec understanding - are you saying that storing mocks as properties should _never_ be done?

**Comment by stof** [comment]
@naomigo exactly.

I even think that you should probably never store anything in a property of the Spec object (btw, doing it would prevent you to spec the behavior of a property with the same name in the subject)

**Comment by naomigo** [comment]
Further to your advice above, through getting rid of the mocked object properties and only calling the mocks when I need them in the tests I have now managed to reduce the run time of the spec from 20s to 10s.

Have learnt a bit along the way too so wil hopefully be able to write more efficient spec files in future!

**Comment by stof** [comment]
yeah, this is the power of relying on the refcount to release memory rather than creating cyclic object graphs and forcing the GC to enter into action, especially when it has to enter into action after each example being run...

**Comment by ciaranmcnulty** [comment]
@stof incidentally I suspect the events we use have a lot of references in them, and there's probably some performance gains there (not to mention logical gains)

**Comment by DonCallisto** [comment]
To all interested, take a look at [this PR](https://github.com/phpspec/prophecy/pull/420)
